"RestDataPanacheException: Failed to list the entities" when table is empty

**Describe the bug**
Using `@RepositoryRestResource` throws an exception when the table is empty.

**Expected behavior**
Call to `/tasks` should return an empty list of tasks, similar to this response with an empty list of countries.

![image](https://user-images.githubusercontent.com/55400451/134951841-f1f840c1-1ef3-43a8-93da-2848c2d44b0b.png)


**Actual behavior**
Call to `/tasks` triggers a `RestDataPanacheException` caused by: `java.lang.IllegalArgumentException: Page index must be >= 0 : -1`.

```
2021-09-27 11:54:18,683 WARN  [io.qua.spr.dat.res.run.RestDataPanacheExceptionMapper] (executor-thread-0) Mapping an unhandled RestDataPanacheException: io.quarkus.rest.data.panache.RestDataPanacheException: Failed to list the entities
	at org.acme.repository.TaskRepositoryJaxRs_47e783e09fc1ec6dc860f6e2c7cfb85a720cc58d.list(TaskRepositoryJaxRs_47e783e09fc1ec6dc860f6e2c7cfb85a720cc58d.zig:671)
	at org.acme.repository.TaskRepositoryJaxRs_47e783e09fc1ec6dc860f6e2c7cfb85a720cc58d_Subclass.list$$superforward1(TaskRepositoryJaxRs_47e783e09fc1ec6dc860f6e2c7cfb85a720cc58d_Subclass.zig:514)
	at org.acme.repository.TaskRepositoryJaxRs_47e783e09fc1ec6dc860f6e2c7cfb85a720cc58d_Subclass$$function$$6.apply(TaskRepositoryJaxRs_47e783e09fc1ec6dc860f6e2c7cfb85a720cc58d_Subclass$$function$$6.zig:57)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:54)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.proceed(InvocationInterceptor.java:62)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor.monitor(InvocationInterceptor.java:49)
	at io.quarkus.arc.runtime.devconsole.InvocationInterceptor_Bean.intercept(InvocationInterceptor_Bean.zig:521)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:41)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:32)
	at org.acme.repository.TaskRepositoryJaxRs_47e783e09fc1ec6dc860f6e2c7cfb85a720cc58d_Subclass.list(TaskRepositoryJaxRs_47e783e09fc1ec6dc860f6e2c7cfb85a720cc58d_Subclass.zig:1115)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130)
	at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:660)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:524)
	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:474)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:476)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:434)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:408)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:69)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
	at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:138)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$1.run(VertxRequestHandler.java:93)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$13.runWith(VertxCoreRecorder.java:543)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.IllegalArgumentException: Page index must be >= 0 : -1
	at io.quarkus.panache.common.Page.<init>(Page.java:53)
	at io.quarkus.panache.common.Page.index(Page.java:118)
	at org.acme.repository.TaskRepositoryJaxRs_47e783e09fc1ec6dc860f6e2c7cfb85a720cc58d.list(TaskRepositoryJaxRs_47e783e09fc1ec6dc860f6e2c7cfb85a720cc58d.zig:376)
```

**To Reproduce**
Steps to reproduce the behavior:
1.  `git clone https://github.com/ethanimproving/empty-list-defect.git`
2. Create a MySQL Database called "scratch" and start the application. Ensure there is NO data in the task table.
3. Send a GET request to `http://localhost:8080/tasks`

**Environment (please complete the following information):**
 - Browser: Google Chrome
 - OS: Microsoft Windows 10 Enterprise

**Additional context**

The exception is handled in `quarkus-spring-data-rest`, so I suspect the issue is with that package.

![image](https://user-images.githubusercontent.com/55400451/134952899-38d08a60-c5b7-4abe-a3d4-fb1152f09c04.png)

Somehow the `list()` method of the JPA implementation of my TaskRepository (`TaskRepositoryJaxRs_47e783e09fc1ec6dc860f6e2c7cfb85a720cc58d`) is calling the `io.quarkus.panache.common` Pagination with a negative index. When my table is empty, my page index is -1 instead of 0.

![image](https://user-images.githubusercontent.com/55400451/134954454-f7f3ee52-2406-4178-bfd6-b8dfdd945cae.png)

--- a/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/AbstractGetMethodTest.java
+++ b/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/AbstractGetMethodTest.java
@@ -328,4 +328,13 @@ public abstract class AbstractGetMethodTest {
             assertThat(link.getRel()).isEqualTo("previous");
         });
     }
+
+    @Test
+    void shouldListEmptyTables() {
+        given().accept("application/hal+json")
+                .and().queryParam("page", 1)
+                .and().queryParam("size", 1)
+                .when().get("/empty-list-items")
+                .then().statusCode(200);
+    }
 }

--- a/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/entity/EmptyListItem.java
+++ b/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/entity/EmptyListItem.java
@@ -1,0 +1,8 @@
+package io.quarkus.hibernate.orm.rest.data.panache.deployment.entity;
+
+import javax.persistence.Entity;
+
+@Entity
+public class EmptyListItem extends AbstractItem<Long> {
+
+}

--- a/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/entity/EmptyListItemsResource.java
+++ b/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/entity/EmptyListItemsResource.java
@@ -1,0 +1,8 @@
+package io.quarkus.hibernate.orm.rest.data.panache.deployment.entity;
+
+import io.quarkus.hibernate.orm.rest.data.panache.PanacheEntityResource;
+import io.quarkus.rest.data.panache.ResourceProperties;
+
+@ResourceProperties(hal = true)
+public interface EmptyListItemsResource extends PanacheEntityResource<EmptyListItem, Long> {
+}

--- a/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/entity/PanacheEntityResourceGetMethodTest.java
+++ b/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/entity/PanacheEntityResourceGetMethodTest.java
@@ -11,7 +11,8 @@ class PanacheEntityResourceGetMethodTest extends AbstractGetMethodTest {
     static final QuarkusUnitTest TEST = new QuarkusUnitTest()
             .withApplicationRoot((jar) -> jar
                     .addClasses(Collection.class, CollectionsResource.class, AbstractEntity.class, AbstractItem.class,
-                            Item.class, ItemsResource.class)
+                            Item.class, ItemsResource.class,
+                            EmptyListItem.class, EmptyListItemsResource.class)
                     .addAsResource("application.properties")
                     .addAsResource("import.sql"));
 }

--- a/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/repository/EmptyListItem.java
+++ b/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/repository/EmptyListItem.java
@@ -1,0 +1,8 @@
+package io.quarkus.hibernate.orm.rest.data.panache.deployment.repository;
+
+import javax.persistence.Entity;
+
+@Entity
+public class EmptyListItem extends AbstractItem<Long> {
+
+}

--- a/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/repository/EmptyListItemsRepository.java
+++ b/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/repository/EmptyListItemsRepository.java
@@ -1,0 +1,9 @@
+package io.quarkus.hibernate.orm.rest.data.panache.deployment.repository;
+
+import javax.enterprise.context.ApplicationScoped;
+
+import io.quarkus.hibernate.orm.panache.PanacheRepository;
+
+@ApplicationScoped
+public class EmptyListItemsRepository implements PanacheRepository<EmptyListItem> {
+}

--- a/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/repository/EmptyListItemsResource.java
+++ b/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/repository/EmptyListItemsResource.java
@@ -1,0 +1,8 @@
+package io.quarkus.hibernate.orm.rest.data.panache.deployment.repository;
+
+import io.quarkus.hibernate.orm.rest.data.panache.PanacheRepositoryResource;
+import io.quarkus.rest.data.panache.ResourceProperties;
+
+@ResourceProperties(hal = true)
+public interface EmptyListItemsResource extends PanacheRepositoryResource<EmptyListItemsRepository, EmptyListItem, Long> {
+}

--- a/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/repository/PanacheRepositoryResourceGetMethodTest.java
+++ b/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/orm/rest/data/panache/deployment/repository/PanacheRepositoryResourceGetMethodTest.java
@@ -12,7 +12,8 @@ class PanacheRepositoryResourceGetMethodTest extends AbstractGetMethodTest {
             .withApplicationRoot((jar) -> jar
                     .addClasses(Collection.class, CollectionsResource.class, CollectionsRepository.class,
                             AbstractEntity.class, AbstractItem.class, Item.class, ItemsResource.class,
-                            ItemsRepository.class)
+                            ItemsRepository.class, EmptyListItem.class, EmptyListItemsRepository.class,
+                            EmptyListItemsResource.class)
                     .addAsResource("application.properties")
                     .addAsResource("import.sql"));
 }

--- a/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/resources/import.sql
+++ b/extensions/panache/hibernate-orm-rest-data-panache/deployment/src/test/resources/import.sql
@@ -3,3 +3,5 @@ insert into collection(id, name) values ('full', 'full collection');
 
 insert into item(id, name, collection_id) values (nextval('hibernate_sequence'), 'first', 'full');
 insert into item(id, name, collection_id) values (nextval('hibernate_sequence'), 'second', 'full');
+
+-- do not add elements to emptylistitem, it should be kept empty

--- a/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/utils/PaginationImplementor.java
+++ b/extensions/panache/rest-data-panache/deployment/src/main/java/io/quarkus/rest/data/panache/deployment/utils/PaginationImplementor.java
@@ -119,7 +119,12 @@ public final class PaginationImplementor {
      */
     private ResultHandle getLastPage(BytecodeCreator creator, ResultHandle page, ResultHandle pageCount) {
         ResultHandle pageNumber = creator.invokeStaticMethod(
-                ofMethod(Integer.class, "sum", int.class, int.class, int.class), pageCount, creator.load(-1));
+                ofMethod(Math.class, "max", int.class, int.class, int.class),
+                creator.invokeStaticMethod(
+                        ofMethod(Integer.class, "sum", int.class, int.class, int.class),
+                        pageCount,
+                        creator.load(-1)),
+                creator.load(0));
         return creator.invokeVirtualMethod(
                 ofMethod(Page.class, "index", Page.class, int.class), page, pageNumber);
     }

--- a/extensions/spring-data-rest/deployment/src/test/java/io/quarkus/spring/data/rest/paged/EmptyListRecord.java
+++ b/extensions/spring-data-rest/deployment/src/test/java/io/quarkus/spring/data/rest/paged/EmptyListRecord.java
@@ -1,0 +1,17 @@
+package io.quarkus.spring.data.rest.paged;
+
+import javax.persistence.Entity;
+
+@Entity
+public class EmptyListRecord extends AbstractEntity<Long> {
+
+    private String name;
+
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+}

--- a/extensions/spring-data-rest/deployment/src/test/java/io/quarkus/spring/data/rest/paged/EmptyListRecordsPagedResourceTest.java
+++ b/extensions/spring-data-rest/deployment/src/test/java/io/quarkus/spring/data/rest/paged/EmptyListRecordsPagedResourceTest.java
@@ -1,0 +1,23 @@
+package io.quarkus.spring.data.rest.paged;
+
+import static io.restassured.RestAssured.given;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+class EmptyListRecordsPagedResourceTest {
+    @RegisterExtension
+    static final QuarkusUnitTest TEST = new QuarkusUnitTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClasses(AbstractEntity.class, EmptyListRecord.class, EmptyListRecordsRepository.class)
+                    .addAsResource("application.properties"));
+
+    @Test
+    void shouldListEmptyTable() {
+        given().accept("application/json")
+                .when().get("/empty-list-records")
+                .then().statusCode(200);
+    }
+}

--- a/extensions/spring-data-rest/deployment/src/test/java/io/quarkus/spring/data/rest/paged/EmptyListRecordsRepository.java
+++ b/extensions/spring-data-rest/deployment/src/test/java/io/quarkus/spring/data/rest/paged/EmptyListRecordsRepository.java
@@ -1,0 +1,6 @@
+package io.quarkus.spring.data.rest.paged;
+
+import org.springframework.data.repository.PagingAndSortingRepository;
+
+public interface EmptyListRecordsRepository extends PagingAndSortingRepository<EmptyListRecord, Long> {
+}

